core/wait/wait.go

```diff
@@ -12,72 +12,77 @@ import (
 
 var RetryHttpErrorStatusCodes = []int{http.StatusBadGateway, http.StatusGatewayTimeout}
 
-type WaitFn func() (res interface{}, done bool, err error)
-
-type Handler struct {
-	fn                WaitFn
+// AsyncActionCheck reports whether a specific async action has finished.
+//   - waitFinished == true if the async action is finished, false otherwise.
+//   - response contains data regarding the current state of the resource targeted by the async action (if applicable). resource != nil if waitFinished == true.
```

Comment:
> resource != nil if waitFinished == true.

Did you mean `resource != nil if (...)`? Maybe we have cases where `resource` is `nil` even though the wait has finished, so we could rephrase it to `response == nil if waitFinished == false`.

Reply:
I did mean `resource != nil if waitFinished == true`.

It's technically possible for the function to not want to return anything when `waitFinished == true`, and the handler's `WaitWithContext` doesn't care if `resp` is `nil` or not. I'd say it's very unlikely that such an use case happens, though: for the function to decide that the wait is over, it must have done by using some API response, and so it can return that API response that proves the wait has finished (no matter if successful or unsuccessful).

I've checked, and not all usages of `AsyncActionHandler` this rule. Eg, `CreateLoadBalancerWaitHandler`, upon finding that the load balancer's creation failed, returns `true, nil, fmt.Errorf(...)`. However, it can return `true, s, fmt.Errorf(...)`, where `s` is the API response that holds the failed state of the API.

I'd suggest enforcing this rule, by 
- Checking if `resp != nil` in `WaitWithContext`
- Removing the `(if applicable)` portion of the comment
- Adding in the documentation of `WaitWithContext` that `resp != nil` if `err != nil`
- Fix usages of `AsyncActionHandler`

Reply:
We can also enforce a weaker rule, that `resource != nil if waitFinished == true && err != nil`.

Reply:
After discussion, will remove the rule and keep it generic